### PR TITLE
Apply ack-able messages

### PR DESCRIPTION
**Pull request by zackliu**
* Change `JoinGroupMessage` and `LeaveGroupMessage`  to ack-able messages
* Handle `AckMessage` in `DispatchMessageAsync`
* Remove all the partition needed message related codes
* Modify tests

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/AckStatus.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+namespace Microsoft.Azure.SignalR
+{
+    internal static class AckStatus
```
Comment (davidfowl) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/AckStatus.cs]: Why are we using http status codes? A simple 1, 2, 3 would work.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 2]
```diff
@@ -115,7 +114,42 @@ public Task WriteAsync(ServiceMessage serviceMessage)
                 throw new AzureSignalRNotConnectedException();
             }
 
-            return Task.WhenAll(routed.Select(s => Connections[s]).Select(s => s.WriteAsync(serviceMessage)));
+            var tasks = new List<Task>();
```
Comment (davidfowl) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: Why is this code suddenly so complicated? What exactly is it trying to do?

File: test/Microsoft.Azure.SignalR.AspNet.Tests/Infrastructure/TestServiceMessageHandler.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ public override Task WriteAsync(string partitionKey, ServiceMessage serviceMessa
                 throw new InvalidOperationException("Not expected to write before tcs is inited");
             }
 
-            return Task.CompletedTask;
+            return Task.FromResult(true);
```
Comment (davidfowl) [inline test/Microsoft.Azure.SignalR.AspNet.Tests/Infrastructure/TestServiceMessageHandler.cs]: This allocates. Change Task<bool> to ValueTask<bool>

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs (hunk 2 above)
Comment (zackliu) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: We just want to ensure one of the operations succeeded or all of them failed. And if failed from exceptions, throw them.

Comment (davidfowl) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: Why isn't WhenAll enough for that?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 4]
```diff
@@ -158,15 +171,19 @@ public Task WriteAsync(ServiceMessage serviceMessage)
...
-            if (string.IsNullOrEmpty(partitionKey))
+            if (!(serviceMessage is IAckableMessage))
             {
-                throw new ArgumentNullException(nameof(partitionKey));
+                throw new ArgumentException($"{nameof(serviceMessage)} is not {nameof(IAckableMessage)}");
             }
 
-            return WriteToPartitionedConnection(partitionKey, serviceMessage);
+            var task = _ackHandler.CreateAck(out var id);
+            ((IAckableMessage) serviceMessage).AckId = id;
```
Comment (davidfowl) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: nit: Spacing looks weird. Remove the space between ) and serviceMessage

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs (hunk 2 above)
Comment (zackliu) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: To reduce latency in some situations. If one endpoint succeed and another endpoint lost the ack message, we have to wait for ack timeout if using WhenAll. The posibility of that will go higher if there're many endpoints.

File: test/Microsoft.Azure.SignalR.AspNet.Tests/Infrastructure/TestServiceMessageHandler.cs (hunk 3 above)
Comment (zackliu) [reply test/Microsoft.Azure.SignalR.AspNet.Tests/Infrastructure/TestServiceMessageHandler.cs]: This just in test. In other situations, we need to allocate a task.

File: src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs [hunk 5]
```diff
@@ -173,9 +173,9 @@ public override Task AddToGroupAsync(string connectionId, string groupName, Canc
                 return Task.CompletedTask;
             }
 
-            var message = new JoinGroupMessage(connectionId, groupName);
-            // Send this message from a fixed service connection, so that message order can be reserved.
-            return _serviceConnectionManager.WriteAsync(groupName, message);
+            var message = new JoinGroupWithAckMessage(connectionId, groupName, -1);
+
+            return _serviceConnectionManager.WriteAckableMessageAsync(message);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs]: We should pass the cancellation token from user.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: Just use WhenAny in a loop then? Or write a helper to encapsulate that logic

File: test/Microsoft.Azure.SignalR.AspNet.Tests/SignalRMessageParserTest.cs [hunk 6]
```diff
@@ -56,7 +56,7 @@ public void TestAddToGroupCommandMessage()
 
             var msgs = parser.GetMessages(message).ToList();
             Assert.Single(msgs);
-            var msg = msgs[0].Message as JoinGroupMessage;
+            var msg = msgs[0].Message as JoinGroupWithAckMessage;
```
Comment (vwxyzh) [inline test/Microsoft.Azure.SignalR.AspNet.Tests/SignalRMessageParserTest.cs]: Assert.IsType

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 7]
```diff
@@ -115,7 +114,42 @@ public Task WriteAsync(ServiceMessage serviceMessage)
...
+
+            var task = await Task.WhenAny(tcs.Task, Task.WhenAll(tasks));
+
+            if (tcs.Task.IsCompleted)
+            {
+                return true;
+            }
+            else
+            {
+                if (task.IsFaulted && task.Exception != null)
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: `await task;`?

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal sealed class AckHandler : IDisposable
+    {
+        private readonly ConcurrentDictionary<int, AckInfo> _acks = new ConcurrentDictionary<int, AckInfo>();
+        private readonly Timer _timer;
+        private readonly TimeSpan _ackInterval = TimeSpan.FromSeconds(5);
+        private volatile int _currentId = 0;
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: remove `volatile`, the only usage is `Interlocked.Increment`.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 9]
```diff
@@ -158,15 +171,19 @@ public Task WriteAsync(ServiceMessage serviceMessage)
             return WriteToRandomAvailableConnection(serviceMessage);
         }
 
-        public Task WriteAsync(string partitionKey, ServiceMessage serviceMessage)
+        public async Task<bool> WriteAckableMessageAsync(ServiceMessage serviceMessage)
         {
-            // If we hit this check, it is a code bug.
-            if (string.IsNullOrEmpty(partitionKey))
+            if (!(serviceMessage is IAckableMessage))
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: `if (!(serviceMessage is IAckableMessage ackMessage))`

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/AckStatus.cs (hunk 1 above)
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/AckStatus.cs]: >class [](start = 20, length = 5)

use enum?

File: src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs [hunk 10]
```diff
@@ -185,9 +185,9 @@ public override Task RemoveFromGroupAsync(string connectionId, string groupName,
                 return Task.CompletedTask;
             }
 
-            var message = new LeaveGroupMessage(connectionId, groupName);
-            // Send this message from a fixed service connection, so that message order can be reserved.
-            return _serviceConnectionManager.WriteAsync(groupName, message);
+            var message = new LeaveGroupWithAckMessage(connectionId, groupName, -1);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs]: > -1 [](start = 79, length = 3)

what `-1` means? Use `enum` for readability?

File: src/Microsoft.Azure.SignalR.AspNet/HubHost/ServiceMessageBus.cs [hunk 11]
```diff
@@ -59,28 +59,28 @@ private async Task WriteMessage(IServiceConnectionContainer connection, AppMessa
...
+                        await connection.WriteAckableMessageAsync(leaveGroupMessage);
                     }
                     finally
                     {
                         _ackHandler.TriggerAck(appMessage.RawMessage.CommandId);
                     }
                     break;
                 case GroupBroadcastDataMessage groupBroadcastMessage:
-                    await connection.WriteAsync(groupBroadcastMessage.GroupName, groupBroadcastMessage);
+                    await connection.WriteAsync(groupBroadcastMessage);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/HubHost/ServiceMessageBus.cs]: >WriteAsync [](start = 37, length = 10)

can move to `default` now

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 12]
```diff
@@ -115,7 +114,42 @@ public Task WriteAsync(ServiceMessage serviceMessage)
                 throw new AzureSignalRNotConnectedException();
             }
 
-            return Task.WhenAll(routed.Select(s => Connections[s]).Select(s => s.WriteAsync(serviceMessage)));
+            var tasks = new List<Task>();
+            var containers = routed.Select(s => Connections[s]);
+            var tcs = new TaskCompletionSource<bool>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs]: >TaskCompletionSource [](start = 26, length = 20)

use async option for tcs

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 13]
```diff
@@ -82,6 +86,15 @@ protected async Task StartCoreAsync(IServiceConnection connection, string target
 
         public abstract Task HandlePingAsync(PingMessage pingMessage);
 
+        public void HandleAck(AckMessage ackMessage)
+        {
+            if (ackMessage.Status == AckStatus.Ok)
+            {
+                _ackHandler.TriggerAck(ackMessage.AckId, true);
+            }
+            _ackHandler.TriggerAck(ackMessage.AckId, false);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: >_ackHandler [](start = 12, length = 11)

missing `else`

File: src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+                    if (_acks.TryRemove(pair.Key, out var ack))
+                    {
+                        // If acks not coming back in time, do not throw an exception
+                        ack.Tcs.TrySetResult(false);
+                    }
+                }
+            }
+        }
+
+        public void Dispose()
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: >Dispose [](start = 20, length = 7)

who to dispose it?

Comment (zackliu) [reply src/Microsoft.Azure.SignalR.Common/Utilities/AckHandler.cs]: Made containerbase disposable to dispose AckHandler and ready for scaling.

**Pull request merged**